crates/sui-core/src/quorum_driver/mod.rs

```diff
@@ -437,121 +601,132 @@ where
+        metrics: Arc<QuorumDriverMetrics>,
+    ) {
+        while let Some(task) = task_receiver.recv().await {
+            // TODO check reconfig process here
+
+            debug!(?task, "Dequeued task");
+            if Instant::now()
+                .checked_duration_since(task.next_retry_after)
+                .is_none()
+            {
+                // Not ready for next attempt yet, re-enqueue
+                let _ = quorum_driver.enqueue_task(task).await;
+                continue;
             }
+            metrics.current_requests_in_flight.dec();
+            let qd = quorum_driver.clone();
+            let metrics_clone = metrics.clone();
+            spawn_monitored_task!(QuorumDriverHandler::process_task(qd, task, metrics_clone));
         }
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use crate::test_utils::init_local_authorities;
-
-    #[tokio::test]
-    async fn test_not_retry_on_object_locked() -> Result<(), anyhow::Error> {
-        let (auth_agg, _, _) = init_local_authorities(4, vec![]).await;
-
-        let quorum_driver_handler = QuorumDriverHandler::new(
-            Arc::new(auth_agg.clone()),
-            QuorumDriverMetrics::new_for_tests(),
-        );
-        let quorum_driver = quorum_driver_handler.clone_quorum_driver();
-        let validity = quorum_driver
-            .authority_aggregator()
-            .load()
-            .committee
-            .validity_threshold();
-
-        assert_eq!(auth_agg.clone_inner_clients().keys().cloned().count(), 4);
-
-        // good stake >= validity, no transaction will be retried, expect Ok(None)
-        assert_eq!(
-            quorum_driver
-                .attempt_conflicting_transactions_maybe(
-                    validity,
-                    &BTreeMap::new(),
-                    &TransactionDigest::random()
-                )
-                .await,
-            Ok(None)
-        );
-        assert_eq!(
-            quorum_driver
-                .attempt_conflicting_transactions_maybe(
-                    validity + 1,
-                    &BTreeMap::new(),
-                    &TransactionDigest::random()
-                )
-                .await,
-            Ok(None)
-        );
-
-        // good stake < validity, but the top transaction total stake < validaty too, no transaction will be retried, expect Ok(None)
-        let conflicting_tx_digests = BTreeMap::from([
-            (TransactionDigest::random(), (vec![], validity - 1)),
-            (TransactionDigest::random(), (vec![], 1)),
-        ]);
-        assert_eq!(
-            quorum_driver
-                .attempt_conflicting_transactions_maybe(
-                    validity - 1,
-                    &conflicting_tx_digests,
-                    &TransactionDigest::random()
-                )
-                .await,
-            Ok(None)
-        );
-
-        Ok(())
+fn convert_to_quorum_driver_error_if_nonretryable(
```

Comment:
Is the plan to extend this function latter? For instance, if the transaction has invalid sig or input it would be non-retriable. 
A TODO?